# Microbiome

This repository is for files, information and code associated with the Aquatic and Riparian Macroinvertebrate Microbiome project run by Elise Millar and Dr. Karen Kidd (McMaster University). This project was a field study with samples collected from the Grand River at sites upstream and downstream of the Kitchener and Waterloo wastewater treatment plants (Ontario, Canada). The digestive gland (mussels) or whole-body (spiders) or ethanol-rinsed whole-body (invertebrates) was collected for microbiome analyses. Extraction and sequencing was completed in the lab of Dr. Mike Surette and the Farncombe Institute (McMaster University). Field sampling was done in collaboration and with the support of Dr. Patty Gillis' lab (Environment and Climate Change Canada) and Dr. Mark Servos' lab (University of Waterloo).

# Invertebrate, Water, and Effluent Samples

"2018-19_MicrobiomeWWTP_Rcode_v1 (Invert-all, Mussel, Spider, Hydro, Perlid, Baetid, Ephem, Hepta)' was used to analyze 16S rRNA bacterial sequences from invertebrates (spiders, mussels, aquatic insects) collected upstream and downstream of wastewater treatment plants on the Grand River. 

"2018-19_WaterMicrobiomeWWTP_Rcode_v1" was used to analyze 16S rRNA bacterial sequences from water filters collected upstream and downstream of wastewater treatment plants on the Grand River. 

"2018-19_EffluentDerivedGenera_Rcode_v1" was used to analyze 16S rRNA bacterial sequences from effluent samples collected directly from wastewater treatment plants on the Grand River. 

Fastq sequences and metadata associated with this projected have been uploaded to NCBI and are linked in the repository description. Files required to run these scripts include:
- metadata: 2018-19_InvertMicrobiomeWWTP_meta_v1.txt
- assigned taxonomy: 2018-19_InvertMicrobiomeWWTP_taxav34_v1.csv
- sequences: 2018-19_InvertMicrobiomeWWTP_seqv34_v1.csv
